datumaro/plugins/ndr.py

```diff
@@ -0,0 +1,260 @@
+        if over_sample not in ("random", "similarity"):
+            raise ValueError("Invalid over_sample")
+        if under_sample not in ("uniform", "inverse"):
+            raise ValueError("Invalid under_sample")
+        
+        self.duplicated_subset = duplicated_subset
+        self._remove(algorithm, working_subset, num_cut, over_sample, under_sample, **kwargs)
+        
+    def _remove(self, algorithm, subset, num_cut=None, over_sample='random', under_sample='uniform',
+        **kwargs):
+        working_subset = self._extractor.get_subset(subset)
+        having_image = []
+        all_imgs = []
+        for item in working_subset:
+            if item.image.has_data:
+                having_image.append(item)
+                all_imgs.append(item.image.data)
+
+        if num_cut and num_cut > len(all_imgs):
+            raise ValueError("The number of images is smaller than the cut you want")
+        
+        if algorithm == "gradient":
+            all_key, fidx, kept_index, key_counter, removed_index_with_sim \
+                = self._gradient_based(all_imgs, **kwargs)
+        # else: #other algorithms will be added later
+           
+        kept_index = self._keep_cut(num_cut, all_key, fidx, kept_index, key_counter, removed_index_with_sim, over_sample, under_sample)
+        self.kept_item_id = set([having_image[ii].id for ii in kept_index])
+
+    def _gradient_based(self, all_imgs, block_shape=(4, 4), hash_dim=32, sim_threshold=0.5):
+        if len(block_shape) != 2:
+            raise ValueError("Invalid block_shape")
+        if block_shape[0] <= 0 or block_shape[1] <= 0:
+            raise ValueError("block_shape should be positive")
+        if sim_threshold <= 0:
+            raise ValueError("sim_threshold should be large than 0")
+        if hash_dim > 3 * block_shape[0] * block_shape[1]:
+            raise ValueError("hash_dim should be smaller than feature shape")
+        if hash_dim <= 0:
+            raise ValueError("hash_dim should be positive")
+        # Caculate gradient
+        all_clr = np.array([self._cgrad_feature(img, out_wh=block_shape) for img in all_imgs])
+
+        # Compute hash keys from all the features
+        all_clr = np.reshape(all_clr, (len(all_imgs), -1))
+        all_key = self._project(all_clr, hash_dim)
+
+        # Remove duplication using hash
+        clr_dict = {}
+        key_counter = {}
+        kept_index = []
+        removed_index_with_similarity = dict()
+
+        fidx = np.random.permutation(np.arange(len(all_imgs)))
+        for ii in fidx:
+            key = all_key[ii]
+            clr = all_clr[ii]
+            if key not in clr_dict:
+                clr_dict[key] = [clr]
+                key_counter[key] = 1
+                kept_index.append(ii)
+                continue
+
+            # Hash collision: compare dot-product based feature similarity
+            max_sim = np.max(np.dot(clr_dict[key], clr) ** 50)
```

Comment:
Please make the use of `50` more clear. It is unclear if it is just a random number, or something with a strong mathematical proof.

Reply:
We do not have any experimental results for that number. The number is set arbitrarily. Then do we need to change it as a parameter?

Reply:
I don't have any complaints against the value, but I'd better make it a named constant to clarify its use.

Reply:
Oh I understand. Okay, I will update